cmd/kubeadm/app/util/runtime/runtime.go

```diff
@@ -128,16 +163,13 @@ func (runtime *CRIRuntime) RemoveContainers(containers []string) error {
 }
 
 // PullImage pulls the image
-func (runtime *CRIRuntime) PullImage(image string) error {
-	var err error
-	var out []byte
+func (runtime *CRIRuntime) PullImage(image string) (err error) {
 	for i := 0; i < constants.PullImageRetry; i++ {
-		out, err = runtime.crictl("pull", image).CombinedOutput()
-		if err == nil {
+		if _, err = runtime.impl.PullImage(context.Background(), runtime.imageService, &runtimeapi.ImageSpec{Image: image}, nil, nil); err == nil {
```

Comment:
should we use defaultContext here as well?

Reply:
No, pulling images will be done in crictl without the default context: https://github.com/kubernetes-sigs/cri-tools/blob/940c58313b754ad853ed8d93941173b11d29b1bb/cmd/crictl/image.go#L650

Cancelling the context could also cancel the image pull, which usually can take longer than 2  seconds.